### PR TITLE
data/d10_1038_s41586_019_1289_x

### PR DESCRIPTION
**Pull request by szalata**
Many thanks for contributing to sfaira!

**PR Checklist**
Please fill in the appropriate checklist below (delete whatever is not relevant). These are the most common things requested on pull requests (PRs).

 - [ ] This comment contains a description of changes (with reason)
 - [ ] Referenced issue is linked
 - [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [ ] Documentation in `docs` is updated
 - [ ] `docs/release-notes.rst` is updated

**Description of changes**
Please state what you've changed and how it might affect the user.

**Technical details**
Please state any technical details such as limitations, reasons for additional dependencies, benchmarks etc. here.

**Additional context**
Add any other context or screenshots here.

### CODE REVIEW

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    development_stage_obs_key: "Batch"
+    disease: 
+    disease_obs_key: 
+    ethnicity: 
+    ethnicity_obs_key: 
+    gm:
+    gm_obs_key:
+    individual:
+    individual_obs_key: 
+    organ: "brain"
```
Comment (le-ander) [inline sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml:63]: we can be a bit more granular for organ as these organoids are patterned dorsal forebrain organoids. I think the most fitting term from the UBERON ontology would be "pallium"

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    organ_obs_key: 
+    organism: "Homo sapiens"
+    organism_obs_key: 
+    sample_source: "3d_culture"
+    sample_source_obs_key: 
+    sex: 
+    sex_obs_key: "Batch"
```
Comment (le-ander) [inline sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml:70]: As with dev_stage, let's use a dictionary here for consistency:
So:
```
development_stage:
	GSE129519_expression_11a.6mon.txt.gz: "male"
	and so on...
```

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_development_stage.tsv [hunk 3]
```diff
@@ -0,0 +1,8 @@
+source	target	target_id
```
Comment (le-ander) [inline sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_development_stage.tsv:1]: you can remove this file as we do the mapping now via a dictionary in the yaml.

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_sex.tsv [hunk 4]
```diff
@@ -0,0 +1,8 @@
+source	target	target_id
```
Comment (le-ander) [inline sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_sex.tsv:1]: you can remove this file as we do the mapping now via a dictionary in the yaml

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_cell_type.tsv [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+Cycling	neural progenitor cell	CL:0011020
+IPCs	progenitor cell	CL:0011026
+IPCs/Immature PNs	progenitor cell	CL:0011026
+Immature CFuPNs	pyramidal neuron	CL:0000598
+Immature CPNs	pyramidal neuron	CL:0000598
+Immature Interneurons	interneuron	CL:0000099
+Immature PNs	pyramidal neuron	CL:0000598
+RG	radial glial cell	CL:0000681
+Unknown	unknown	unknown
+Ventral Precursors	precursor cell	CL:0011115
```
Comment (le-ander) [inline sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_cell_type.tsv:15]: I'd map ventral precursor to "neural progenitor cell"

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_cell_type.tsv [hunk 6]
```diff
@@ -0,0 +1,17 @@
+source	target	target_id
+Astroglia	astrocyte	CL:0000127
+CFuPNs	pyramidal neuron	CL:0000598
+CPNs	pyramidal neuron	CL:0000598
+Cajal-Retzius	Cajal-Retzius cell	CL:0000695
+Cycling	neural progenitor cell	CL:0011020
+IPCs	progenitor cell	CL:0011026
```
Comment (le-ander) [inline sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_cell_type.tsv:7]: I'd map IPC to "neural progenitor cell"

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_cell_type.tsv [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+IPCs/Immature PNs	progenitor cell	CL:0011026
```
Comment (le-ander) [inline sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_cell_type.tsv:8]: I'd map IPCs/Immature PNs to "neural progenitor cell"

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    fn = os.path.join(data_dir, sample_fn)
+    fn_meta = os.path.join(data_dir, "meta_combined.txt")
+
+    expr_df = pd.read_csv(fn, sep="\t", compression='gzip', index_col=0)
+    expr_df = expr_df.apply(np.expm1)
+    expr_df = expr_df.T
+
+    metadata_df = pd.read_csv(fn_meta, sep="\t", index_col=0, header=0)
+    metadata_df = metadata_df.tail(-1)
+
```
Comment (le-ander) [inline sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.py:18]: Let's add an extra line to specifically store organoid age in days as well:
```python
metadata_df["organoid_age_days"] = pd.Series([{"3mon": "90", "6mon": "180"}[i] for i in metadata_df["Batch"].str.split("_").str[1]], dtype="category")
```

**Review by le-ander — COMMENTED** [review]
Thanks so much for posting this dataloader, Artur!
I've suggested some changes in the code.

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        - "GSE129519_expression_PGP1.3mon.batch2.txt.gz"
+        - "GSE129519_expression_PGP1.3mon.txt.gz"
+        - "GSE129519_expression_PGP1.6mon.batch3.txt.gz"
+        - "GSE129519_expression_PGP1.6mon.txt.gz"
+dataset_wise:
+    author: "Velasco, Silvia"
+    default_embedding: 
+    doi_preprint: 
+    doi_journal: "10.1038/s41586-019-1289-x"
+    download_url_data:
```
[the remark this answers is not in the document]
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml:16]: updated

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+dataset_or_observation_wise:
+    assay_sc: "10x 3' v2"
+    assay_sc_obs_key: 
+    assay_differentiation: 
+    assay_differentiation_obs_key: 
+    assay_type_differentiation: 
+    assay_type_differentiation_obs_key: 
+    bio_sample:
+    bio_sample_obs_key: 
+    cell_line:
```
[the remark this answers is not in the document]
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml]: updated

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    feature_reference: "Homo_sapiens.GRCh38.84"
+    feature_reference_var_key: 
+    feature_type: "rna"
+    feature_type_var_key: 
+dataset_or_observation_wise:
+    assay_sc: "10x 3' v2"
+    assay_sc_obs_key: 
+    assay_differentiation: 
+    assay_differentiation_obs_key: 
+    assay_type_differentiation:
```
[the remark this answers is not in the document]
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml]: added

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    layer_velocity: 
+dataset_or_feature_wise:
+    feature_reference: "Homo_sapiens.GRCh38.84"
+    feature_reference_var_key: 
+    feature_type: "rna"
+    feature_type_var_key: 
+dataset_or_observation_wise:
+    assay_sc: "10x 3' v2"
+    assay_sc_obs_key: 
+    assay_differentiation:
```
[the remark this answers is not in the document]
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml]: added

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    feature_type_var_key: 
+dataset_or_observation_wise:
+    assay_sc: "10x 3' v2"
+    assay_sc_obs_key: 
+    assay_differentiation: 
+    assay_differentiation_obs_key: 
+    assay_type_differentiation: 
+    assay_type_differentiation_obs_key: 
+    bio_sample:
+    bio_sample_obs_key:
```
[the remark this answers is not in the document]
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml]: added

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_sex.tsv (hunk 4 above)
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_sex.tsv]: done

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_cell_type.tsv (hunk 5 above)
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_cell_type.tsv:15]: done

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_cell_type.tsv (hunk 6 above)
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_cell_type.tsv:7]: done

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_cell_type.tsv (hunk 7 above)
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_cell_type.tsv:8]: done

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.py (hunk 8 above)
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.py:18]: done

**Comment by szalata** [comment]
all the comments should be addressed now

**Comment by le-ander** [comment]
Thanks a lot, Artur. this already looks great. Could you adress the two outstanding comments in the yaml and the one reagrding the dev_stage tsv file?

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    assay_type_differentiation: 
+    assay_type_differentiation_obs_key: 
+    bio_sample:
+    bio_sample_obs_key: 
+    cell_line: 
+    cell_line_obs_key: 
+    cell_type: 
+    cell_type_obs_key: "CellType"
+    development_stage: 
+    development_stage_obs_key: "Batch"
```
[the remark this answers is not in the document]
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml]: done

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml (hunk 1 above)
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001.yaml]: done

File: sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_development_stage.tsv (hunk 3 above)
Comment (szalata) [reply sfaira/data/dataloaders/loaders/d10_1038_s41586_019_1289_x/homosapiens_brain_2019_10x3v2_velasco_001_development_stage.tsv]: done

**Comment by szalata** [comment]
hopefully it's ok now

**Comment by le-ander** [comment]
Looks great, thanks so much! 🚀

**Review by le-ander — APPROVED** [review]

**Pull request merged**